core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -416,24 +419,162 @@ public ShardRouting initialize(ShardRouting shard, String nodeId, @Nullable Stri
      *
      * @return pair of source relocating and target initializing shards.
      */
-    public Tuple<ShardRouting,ShardRouting> relocate(ShardRouting shard, String nodeId, long expectedShardSize) {
+    public Tuple<ShardRouting,ShardRouting> relocateShard(ShardRouting startedShard, String nodeId, long expectedShardSize) {
         ensureMutable();
         relocatingShards++;
-        ShardRouting source = shard.relocate(nodeId, expectedShardSize);
+        ShardRouting source = startedShard.relocate(nodeId, expectedShardSize);
         ShardRouting target = source.getTargetRelocatingShard();
-        updateAssigned(shard, source);
+        updateAssigned(startedShard, source);
         node(target.currentNodeId()).add(target);
         assignedShardsAdd(target);
         addRecovery(target);
         return Tuple.tuple(source, target);
     }
 
     /**
+     * Applies the relevant logic to start an initializing shard.
+     *
+     * Moves the initializing shard to started. If the shard is a relocation target, also removes the relocation source.
+     *
+     * @return the started shard
+     */
+    public ShardRouting startShard(ESLogger logger, ShardRouting initializingShard) {
+        logger.trace("{} marked shard as started (routing: {})", initializingShard.shardId(), initializingShard);
```

Comment:
The logging line should probably go after the `started(initializingShard)` call, otherwise it's not accurate because the shard could be invalid and end up not started